aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/AwsPresigningDecorator.kt

```diff
@@ -0,0 +1,174 @@
+import software.amazon.smithy.rustsdk.traits.PresignableTrait
+
+private val PRESIGNABLE_OPERATIONS = listOf(
+    // TODO(PresignedReqPrototype): Add the other presignable operations
+    ShapeId.from("com.amazonaws.s3#GetObject"),
+    ShapeId.from("com.amazonaws.s3#PutObject"),
+)
+
+// TODO(PresignedReqPrototype): Write unit test
+class AwsPresigningDecorator : RustCodegenDecorator {
+    companion object {
+        val ORDER: Byte = 0
+    }
+
+    override val name: String = "AwsPresigning"
+    override val order: Byte = ORDER
+
+    override fun extras(protocolConfig: ProtocolConfig, rustCrate: RustCrate) {
+        rustCrate.mergeFeature(Feature("client", default = true, listOf("tower")))
+    }
+
+    override fun operationCustomizations(
+        protocolConfig: ProtocolConfig,
+        operation: OperationShape,
+        baseCustomizations: List<OperationCustomization>
+    ): List<OperationCustomization> = baseCustomizations + listOf(
+        AwsInputPresignedMethod(protocolConfig.runtimeConfig, protocolConfig.symbolProvider, operation)
+    )
+
+    /** Adds presignable trait to known presignable operations */
+    override fun transformModel(service: ServiceShape, model: Model): Model {
+        return ModelTransformer.create().mapShapes(model) { shape ->
+            if (shape is OperationShape && PRESIGNABLE_OPERATIONS.contains(shape.id)) {
+                shape.toBuilder().addTrait(PresignableTrait()).build()
+            } else {
+                shape
+            }
+        }
+    }
+}
+
+class AwsInputPresignedMethod(
+    runtimeConfig: RuntimeConfig,
+    private val symbolProvider: RustSymbolProvider,
+    private val operationShape: OperationShape
+) : OperationCustomization() {
+    private val codegenScope = arrayOf(
+        "aws_hyper" to runtimeConfig.awsRuntimeDependency("aws-hyper").copy(optional = true).asType(),
+        "Error" to AwsRuntimeType.Presigning.member("config::Error"),
+        "PresignedRequest" to AwsRuntimeType.Presigning.member("request::PresignedRequest"),
+        "PresignedRequestService" to AwsRuntimeType.Presigning.member("service::PresignedRequestService"),
+        "PresigningConfig" to AwsRuntimeType.Presigning.member("config::PresigningConfig"),
+        "SdkError" to CargoDependency.SmithyHttp(runtimeConfig).asType().member("result::SdkError"),
+        "sig_auth" to runtimeConfig.sigAuth().asType(),
+        "tower" to CargoDependency.Tower.asType(),
+    )
+
+    override fun section(section: OperationSection): Writable = writable {
+        if (section is OperationSection.InputImpl && section.operationShape.hasTrait(PresignableTrait::class.java)) {
+            writeInputPresignedMethod()
+        } else {
+            // TODO(PresignedReqPrototype): Is there a better way to do this?
+            // HACK: Add tower to optional dependencies so that the client feature can reference it
+            rustTemplate("// ignore-me: #{tower}", *codegenScope)
```

Comment:
ohhhh, I see. You can manually call `addDependency`. 

But I admittedly, I'm confused: we shouldn't be adding the feature unless we need presigners. So if we don't generate any presigners, we shouldn't have that feature.

You could also skip the tower dep and use `map_middleware` I think? You might just need to unwrap the SdkSuccess/SdkError?